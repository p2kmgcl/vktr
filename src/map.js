import isFunction from "./assertions/is-function";
import isVector from "./assertions/is-vector";
import Vector from "./vector";

/**
 * Returns a new vector result of applying the
 * given function to each component of the given vector.
 * @param {Vector} a
 * @param {Function} fn
 * @return {Vector}
 */
function map(a, fn) {
  isFunction(fn);
  isVector(a);

  return new Vector(...a.components.map(fn));
}

export default map;
