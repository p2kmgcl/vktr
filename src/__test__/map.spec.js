import map from "../map";
import Vector from "../vector";

describe("map()", () => {
  it("applies the given function to each component", () => {
    const by2 = jest.fn(a => a * 2);
    const a = new Vector(1, 2, 3);
    const b = map(a, by2);

    expect(b).toMatchSnapshot();
    expect(by2.mock.calls).toMatchSnapshot();
  });

  it("requires a function parameter", () => {
    const a = new Vector(1);

    expect(() => map(a, null)).toThrow(TypeError);
    expect(() => map(a, "a")).toThrow(TypeError);
    expect(() => map(a, [1])).toThrow(TypeError);
    expect(() => map(a, NaN)).toThrow(TypeError);
  });
});
