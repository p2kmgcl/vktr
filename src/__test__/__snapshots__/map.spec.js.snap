// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map() applies the given function to each component 1`] = `
Array [
  2,
  4,
  6,
]
`;

exports[`map() applies the given function to each component 2`] = `
Array [
  Array [
    1,
    0,
    Array [
      1,
      2,
      3,
    ],
  ],
  Array [
    2,
    1,
    Array [
      1,
      2,
      3,
    ],
  ],
  Array [
    3,
    2,
    Array [
      1,
      2,
      3,
    ],
  ],
]
`;
